lambda/matching.ml

```diff
@@ -313,64 +313,147 @@ end = struct
 
-exception OrPat
+let flatten_matrix size pss =
+  List.fold_right
+    (fun ps r ->
+      match ps with
+      | [ p ] -> flatten_pat_line size p r
+      | _ -> fatal_error "Matching.flatten_matrix")
+    pss []
 
-let specialize_matrix matcher pss =
-  let rec filter_rec = function
-    | (p :: ps) :: rem -> (
-        match p.pat_desc with
-        | Tpat_alias (p, _, _) -> filter_rec ((p :: ps) :: rem)
-        | Tpat_var _ -> filter_rec ((omega :: ps) :: rem)
-        | _ -> (
-            let rem = filter_rec rem in
-            try matcher p ps :: rem with
-            | NoMatch -> rem
-            | OrPat -> (
-                match p.pat_desc with
-                | Tpat_or (p1, p2, _) ->
-                    filter_rec [ p1 :: ps; p2 :: ps ] @ rem
-                | _ -> assert false
-              )
-          )
-      )
-    | [] -> []
-    | _ ->
-        pretty_matrix Format.err_formatter pss;
-        fatal_error "Matching.specialize_matrix"
-  in
-  filter_rec pss
+(** A default environment (referred to as "reachable trap handlers" in the
+    paper), is an ordered list of [matrix * raise_num] pairs, and are used to
+    decide where to jump next if none of the rows in a given matrix match the
+    input.
+
+    In such situations, one thing you can do is to jump to the first (leftmost)
+    [raise_num] in that list (by doing a raise to the static-cach handler number
+    [raise_num]); and you can assume that if the associated pm doesn't match
+    either, it will do the same thing, etc.
+    This is what [mk_failaction_neg] (and its callers) does.
+
+    A more sophisticated alternative is to use what you know about the input
+    (what you might already have matched) and the current pm (what you know you
+    can't match) to directly jump to a pm that might match it instead of the
+    next one; that is why we don't just keep [raise_num]s but also the
+    associated matrices.
+    [mk_failaction_pos] does (a slightly more sophisticated version of) this.
+*)
+module Default_environment : sig
+  type t
 
-let specialize_default matcher env =
-  let rec make_rec = function
-    | [] -> []
-    | ([ [] ], i) :: _ -> [ ([ [] ], i) ]
-    | (pss, i) :: rem -> (
-        let rem = make_rec rem in
-        match specialize_matrix matcher pss with
-        | [] -> rem
-        | [] :: _ -> [ ([ [] ], i) ]
-        | pss -> (pss, i) :: rem
-      )
-  in
-  make_rec env
+  val is_empty : t -> bool
 
-let default_pop_column def = specialize_default (fun _p rem -> rem) def
+  val pop : t -> ((matrix * int) * t) option
 
-let default_pop_compat p def =
-  let compat_matcher q rem =
-    if may_compat p q then
-      rem
-    else
-      raise NoMatch
-  in
-  specialize_default compat_matcher def
+  val empty : t
+
+  val cons : matrix -> int -> t -> t
+
+  val specialize : (pattern -> pattern list -> pattern list) -> t -> t
+
+  val pop_column : t -> t
+
+  val pop_compat : pattern -> t -> t
```

Comment:
Isn't there a contradiction between the `pop_column` function, which remove one column from every matrix, and `pop_compat`, which remove one column from compatible rows but also remove incompatible rows? Maybe keeping `specialize_compat` as a name would be clearer here.

Reply:
Yes, I kept the names that were already there, but I guess just `specialize` would be a decent name.

There is also `val pop : ...` which pops the first matrix from the list... Perhaps `pop_column` should become `pop_first_column_of_matrices`?

Reply:
Another possibility might be `specialize_for_var` to describe the use case rather than the implementation?

Reply:
`pop_compat` is the composition of `pop_column` and `filter_compat` (or `guard_compat`). I don't think that `specialize` is more informative as a name, and I'm not fond of the longer names proposed (`pop_first_column_of_matrices` should be `map_rows pop`); maybe just implement `filter_compat` and write the composition by hand? (It's one more traversal of the whatever, but it doesn't actually matter much.)

Reply:
@trefis and myself have discussed better interfaces for matchers, in particular `pattern -> pattern list` (where NoMatch is the empty list and OrPat returns a list of explosions) and a version where `OrPat` is removed (replaced by generic handling of or-patterns in the specialization code), allowing `pattern -> pattern option`. I think that at this point it may be possible to keep only `specialize` in the module, and write the matchers nicely in the client code, obviating the need for those API naming discussions.

In the meantime, my inclination would be to keep the names as they currently are to keep the refactoring train going.

Reply:
It is most probably a good idea to let the refactoring train go on. And, maybe, or not, I will have a look at the end of the refactorization tunnel to see if those names are still here.